### DISCUSSION TITLE
a way for clients and customers to leave reviews

### DISCUSSION THREAD

**Discussion by Goldmaster** [squidfunk/mkdocs-material#7354] [Q&A]
I am looking at moving away from wordpress to mkdocs due to more simplicity of markdown. However, I am wondering if there is such feature or plugin that allows clients and customers to leave reviews and feedback similar to wp customer reviews wordpress plugin.

If not then a subdomain is OK but then I'm not sure what could be used other than wordpress with wp customer reviews.

Many thanks for your help.

**Comment by kamilkrzyskow** [comment]
Hello @Goldmaster,
MkDocs is only a static site generator, serving HTML after processing Markdown. Wordpress is a CMS built with PHP, and by design it requires a database server to store data, and wordpress plugins use that database to add their own data, and load it after a page refresh.

If you're looking for "a way" there are different options with different degrees of required tech know-how to be able to customize it:
1. You could use the comments functionality:
    - https://squidfunk.github.io/mkdocs-material/setup/adding-a-comment-system/
    
    for people to add their input, this would require some external account for Giscus or Disqus etc. and those are normal comments, so there is no starring system.

2. Use a "Contact Me" form service as a wrapper for getting Reviews and use some email API to periodically fetch the email in CI using a cron job, and rebuild the static mkdocs page with new reviews:
    - For example to add a contact form backend use un-static\.com there are many other similar services, most of them have free plans with monthly message limits.
    - Here is an example with the above https://github.com/squidfunk/mkdocs-material/discussions/6746
    - I have no reference for the email API handling and mergin it with CI
    
    this approach would allow you to add a rating/star system but as you'd fetch the emails periodically using the some API the reviews won't appear on the page straight away.
    
3. Find a similar service to both 1. and 2. but instead of Contact Form or Comments it would provide a form to gather ratings in real time.

4. If you're going to keep using your old server with the PHP then you could add a form to send data to your Wordpress Reviews Plugin php page, and it would add the reviews the same as before. However, fetching them would require a similar cron job-based approach or to get less latency you could modify the plugin and from your backend server invoke a rebuild of the mkdocs website. I can see that the wordpress review plugin has some Javascript included, so perhaps some of it can be reused in Mkdocs.

Reference:
- https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form
- https://www.w3schools.com/tags/ref_httpmethods.asp

**Reply by Goldmaster** [reply]
Many thanks for your help and advice.

I was going to ask about a contact form using hcaptcha.

One option I could have a look into is seeing if there is a self hosted review system. I could then run it on a subdomain or try and get it onto the same url for example domain.com/reviews rather than reviews.domain.com 

Another thought is to setup wordpress on a subdomain just for wp reviews and contact form 7 but it defeats the object and would may as well stick with wordpress.

Essentially, some form of a database system is needed to keep track of reviews or I just manually add them myself.

**Reply by kamilkrzyskow** [reply]
> Essentially, some form of a database system is needed to keep track of reviews or I just manually add them myself.

Yes, that's the gist of it.

> I was going to ask about a contact form using hcaptcha.

You'll likely have to research the market on your own, use sites like alternative.to to find similar services to the one provided previously.

> One option I could have a look into is seeing if there is a self hosted review system. I could then run it on a subdomain or try and get it onto the same url for example domain.com/reviews rather than reviews.domain.com

A valid idea. I haven't proposed it before, as this could entail quite a lot of work compared to the other "battle-tested" solutions, as this can be tricky to do correctly. In your local implementation, even if you add some Captcha, there might still be some malicious actors that could send advertisement / swear words in the body of the review. I would assume other services have some (at least minimal) protection / safeguards in-place for this occasion.

 Assuming you have control over the backend, serving the files, you could easily setup a rewrite rule for the /reviews path.
 
 > Another thought is to setup wordpress on a subdomain just for wp reviews and contact form 7 but it defeats the object and would may as well stick with wordpress.
 
Like I said before, you could likely extract a lot of the code out of the plugin and use it in that /reviews php endpoint to handle form requests. Though, I'm not really sure what's so great about this feedback plugin, so I might have a superficial view on things.
